### PR TITLE
LOB-196-event-notifications-flows-v4

### PR DESCRIPTION
**Pull request by alinolteanufr**
- Adding Controllers and Unit Tests for event notifications flow v4

### CODE REVIEW

**Review by wmorrison365fr — APPROVED** [review]
Approved, on the basis it's just to deply and test (not actually merged in this state).

File: secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/aggregatedpolling/AggregatedPollingApiController.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+import static com.forgerock.sapi.gateway.ob.uk.common.datamodel.converter.v4.event.FREventPollingConverter.toFREventPolling;
+
+@Controller("AggregatedPollingApiV4.0.0")
+@Slf4j
+public class AggregatedPollingApiController implements AggregatedPollingApi {
+    private final EventPollingService eventPollingService;
+
+    public AggregatedPollingApiController(EventPollingService eventPollingService) {
+        this.eventPollingService = eventPollingService;
```
Comment (wmorrison365fr) [inline secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/aggregatedpolling/AggregatedPollingApiController.java:38]: Null check?

File: secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/aggregatedpolling/AggregatedPollingApiController.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright © 2020-2024 ForgeRock AS (obst@forgerock.com)
```
Comment (wmorrison365fr) [inline secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/aggregatedpolling/AggregatedPollingApiController.java:2]: Again, just:
```suggestion
 * Copyright © 2024 ForgeRock AS (obst@forgerock.com)
```
?

File: secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/eventsubscription/EventSubscriptionsApiController.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+    public EventSubscriptionsApiController(EventSubscriptionsRepository eventSubscriptionsRepository) {
+        this.eventSubscriptionsRepository = eventSubscriptionsRepository;
+    }
+
+    @Override
+    public ResponseEntity changeEventSubscriptionsEventSubscriptionId(String eventSubscriptionId, String authorization, String xJwsSignature, OBEventSubscriptionResponse1 obEventSubscriptionResponse1, String xFapiAuthDate, String xFapiCustomerIpAddress, String xFapiInteractionId, String xCustomerUserAgent, String tppId, HttpServletRequest request) throws OBErrorResponseException {
+        OBEventSubscription1 updatedSubscription =
+                new OBEventSubscription1().data(new OBEventSubscription1Data()
+                        .callbackUrl(obEventSubscriptionResponse1.getData().getCallbackUrl())
```
Comment (wmorrison365fr) [inline secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/eventsubscription/EventSubscriptionsApiController.java:60]: Can `obEventSubscriptionResponse1` be null here?

File: secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/eventsubscription/EventSubscriptionsApiController.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+            FREventSubscription existingEventSubscription = byId.get();
+            // A TPP must not update an event-subscription on an older version, via the EventSubscriptionId for an event-subscription created in a newer version
+            OBVersion apiVersion = VersionPathExtractor.getVersionFromPath(request);
+            OBVersion resourceVersion = OBVersion.fromString(existingEventSubscription.getEventSubscription().getVersion());
+            if (ResourceVersionValidator.isAccessToResourceAllowed(apiVersion, resourceVersion)) {
+                existingEventSubscription.setEventSubscription(toFREventSubscriptionData(updatedSubscription));
+                eventSubscriptionsRepository.save(existingEventSubscription);
+                return ResponseEntity.ok(packageResponse(existingEventSubscription));
+            } else {
+                return ResponseEntity.status(HttpStatus.CONFLICT).body("The event subscription can't be update via an older API version.");
```
Comment (wmorrison365fr) [inline secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/eventsubscription/EventSubscriptionsApiController.java:75]: s/update/updated
```suggestion
                return ResponseEntity.status(HttpStatus.CONFLICT).body("The event subscription can't be updated via an older API version.");
```

File: secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/eventsubscription/EventSubscriptionsApiController.java [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+    @Override
+    public ResponseEntity changeEventSubscriptionsEventSubscriptionId(String eventSubscriptionId, String authorization, String xJwsSignature, OBEventSubscriptionResponse1 obEventSubscriptionResponse1, String xFapiAuthDate, String xFapiCustomerIpAddress, String xFapiInteractionId, String xCustomerUserAgent, String tppId, HttpServletRequest request) throws OBErrorResponseException {
+        OBEventSubscription1 updatedSubscription =
+                new OBEventSubscription1().data(new OBEventSubscription1Data()
+                        .callbackUrl(obEventSubscriptionResponse1.getData().getCallbackUrl())
+                        .eventTypes(obEventSubscriptionResponse1.getData().getEventTypes())
+                        .version(obEventSubscriptionResponse1.getData().getVersion())
+                );
+        Optional<FREventSubscription> byId = eventSubscriptionsRepository.findById(eventSubscriptionId);
```
Comment (wmorrison365fr) [inline secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/eventsubscription/EventSubscriptionsApiController.java:64]: Can this be null?

File: secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/eventsubscription/EventSubscriptionsApiController.java
Comment (wmorrison365fr) [inline secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/eventsubscription/EventSubscriptionsApiController.java:1]: Ok, I see... these classes are lifted from v3.1, and pushed to separate v4.0 spec package.

Comment (alinolteanufr) [reply secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/eventsubscription/EventSubscriptionsApiController.java:1]: These classes are lifted from v3.1.10 to v4.0.0 with specific v4.0.0 datamodel from common repo

File: secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/eventsubscription/EventSubscriptionsApiController.java [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+import com.forgerock.sapi.gateway.uk.common.shared.api.meta.obie.OBVersion;
+import jakarta.servlet.http.HttpServletRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Controller;
+import uk.org.openbanking.datamodel.v4.common.Links;
+import uk.org.openbanking.datamodel.v4.common.Meta;
+import uk.org.openbanking.datamodel.v4.event.*;
```
Comment (wmorrison365fr) [inline secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/eventsubscription/EventSubscriptionsApiController.java:37]: Can you import specific classes?

Also line 40?

File: secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/eventsubscription/EventSubscriptionsApiController.java
Comment (wmorrison365fr) [reply secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/eventsubscription/EventSubscriptionsApiController.java:1]: Doing that comparison, with order etc reverted, there's only really method name changes and ordering in this class? 

Would be quick to review without these things.

Are the method name changes needed?

**Review by wmorrison365fr — APPROVED** [review]

File: secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/eventsubscription/EventSubscriptionsApiController.java
Comment (alinolteanufr) [reply secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/eventsubscription/EventSubscriptionsApiController.java:1]: The interfaces were generated by Dan, I just added the missing parameters without changing the methods name that he used. In the controllers, things that are changed in comparison to v3.1.10 are the OB objects to match v4, the interface used and the methods from the converters.

Comment (wmorrison365fr) [reply secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v4_0_0/eventsubscription/EventSubscriptionsApiController.java:1]: Looking further into this, the interface are also completely different:
- [EventSubscriptionsApi (3.1.10)](https://github.com/SecureApiGateway/secure-api-gateway-ob-uk-rs/blob/db9539f484976c22f14d1d5ae89135e1c6215f2d/secure-api-gateway-ob-uk-rs-obie-api/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/obie/api/event/v3_1_10/eventsubscription/EventSubscriptionsApi.java#L45)
- [EventSubscriptionsApi (4.0.0)](https://github.com/SecureApiGateway/secure-api-gateway-ob-uk-rs/blob/db9539f484976c22f14d1d5ae89135e1c6215f2d/secure-api-gateway-ob-uk-rs-obie-api/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/obie/api/event/v4_0_0/eventsubscription/EventSubscriptionsApi.java#L52)

So, the implementations are very different between this (4.0) and the [3.1.10](https://github.com/SecureApiGateway/secure-api-gateway-ob-uk-rs/blob/ob-v4/secure-api-gateway-ob-uk-rs-server/src/main/java/com/forgerock/sapi/gateway/ob/uk/rs/server/api/obie/event/v3_1_10/eventsubscription/EventSubscriptionsApiController.java#L60) implementation.

The interfaces are generated though, so this is out of our control. Creates a lot of noise though, which then ripples out.

**Review by wmorrison365fr — APPROVED** [review]

**Pull request merged**